### DISCUSSION TITLE
General usage question

### DISCUSSION THREAD

**Discussion by MostHated** [quabug/OneShot#21] [General]
Hey there,
    Pardon my overall IoC/DI ignorance, but I had a question about usage of this system. 

In the examples in the readme, you have the creation of the container class as seen below:

```cs
var container = new Container();
```
This container, is this supposed to be, for lack of a better term, the "main" container in which the rest of your project should be registering and obtaining dependencies? If this is the case, where is it supposed to actually be constructed, on a singleton MonoBehaviour in scene, or should a container be created in each class that you plan to utilize the DI functionality? It is a bit hard to practically tell exactly where/how this is supposed to be used without a tiny bit of additional context. 

 I spend a lot of time making editor tools and I wanted to test out if using DI might help streamline my workflow, so I figured a system that seemingly is much less complex overall than something like Zenject might be a good place to start, but I am embarrassed to say, even with this I am rather unsure of where exactly to start. Would the ideal entrypoint end up being on a ScriptableObject in which all other classes in the project have access (ex. A ScriptableObject class which creates a public  container object, along with the OneShot script within their own assembly definition within my project, then have all other assemblies reference it?)

Do you perhaps have a tiny and very basic example project with just a few items that I might be able to look at, to see where and how things are setup? Just so that I could have some sort of a 'frame of reference' on how I could start out? If not, that is perfectly fine, I completely understand.

Thanks!
-MH

**Comment by quabug** [comment]
Hi @MostHated,
Thank you for asking this question. I am also thinking about publishing my unity component at some point.
I highly recommend reading documents of `Zenject` or `vContainer` to get familiar with automatic DI.

About your question, let me go back a bit to the basic concepts of `Container`.
1. `Container` is just a map of `Type => Func<object>` or `Type => () => object`
2. `Container.Register` is the process to build this map.
3. `Container.Resolve` is just like calling function `container(type)()`, while `Inject` is several `Resolve`s on parameter types.

And the typical process to resolve a type or inject parameter types is like:
1. Create a `Container`.
2. `Register` types into `Container`.
3. `Resolve` types or `Inject` types into the **instance**.

It might be easier to understand how to achieve DI in unity by keeping these fundamental concepts in mind.

Now let's create some components to achieve automatic DI for Unity components:
``` C#
class MyContainer : MonoBehavior
{
  public Container Value = new Container();
}

class IntRegister : MonoBehavior
{
  public MyContainer Container;
  void Awake() => Container.Value.RegisterInstance(123).AsSelf();
}

class Injector : MonoBehavior
{
  public MyContainer Container;
  void Start() => Container.Value.InjectAll(GetComponent<InjectInt>());
}

class InjectInt : MonoBehavior
{
  [Inject] void Inject(int value) => Debug.Log(value);
}
```
With those components, you can inject an `int` into the `InjectInt` component by creating a container, registering `int` instance into the container, and injecting `int` into the component.

And I also create [a sample of `RecursiveInjector`](https://github.com/quabug/OneShot/tree/unity/Assets/RecursiveInjection) which able to create container hierarchy by gameobjects' hierarchy, and [a sample of `ContainerScriptableObject`](https://github.com/quabug/OneShot/tree/unity/Assets/ScriptableObjectContainer) which act like a global container.

I also encourage you to create your component for your needs if necessary. Just find your way to *create*, *register* and *inject*.

quabug

**Comment by MostHated** [comment]
I thank you much for this. That example branch is perfect, as it allows me to follow the live working callstack and step through the methods as they are called. With most things, I don't usually need that in-depth of an analysis to understand what is going on, but for some reason, I needed it with this. What you provided did the trick, though, so now I should be good to go. :+1: 

Thank you again, I very much appreciate it.
-MH

**Comment by MostHated** [comment]
So, building off of your ScriptableObject example, I wanted to make an "all-in-one" type of registration and injection handler that didn't rely heavily on the scene, and what I have worked up so far seems to be working out well. I added two additional attributes. The first is 'DIInitializer', which is a method attribute that gets called by an initializer in order to handle all of the registration to the container. The second is 'Injectable', which is a class attribute that I simply put on the classes in which I planned to do injection so that I could quickly and easily find them and run InjectAll on just the specific classes that needed it. 

All of my tools and objects I have tested out so far are able to be injected automatically at edit time as I was hoping. So I am pleased with that. The link below should be the majority of what I had put together. I need to clean it up a bit and test it in a build so I can make sure things are fired up how I am expecting (awake, enable, runtimeinitializeonload, etc) and remove some calls to a base class and a few other things, but it's enough to get the overall idea. I just wanted to share it in case it helps anyone in the future, or if there is any feedback on it.

Thanks again for the help the other day. As I mentioned, it was exactly what I needed to get me going. :+1:

https://gist.github.com/instance-id/e98ddf640e71449e84109f72fffa5fd9

<img src="https://img.instance.id/86T1JfC142SZ/direct" width="600px">
